docker_swarm_service: Fix crash when using configs parameter

hannseman opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
Fixes #50467

`DockerService.configs` can safely initially be a list. When used it is checked for a boolean value and thus both `None` and `[]` will return `False`.

Examples:
https://github.com/hannseman/ansible/blob/4bfc254ac4dbf31afb0e5fc3d6c11c94bad9785e/lib/ansible/modules/cloud/docker/docker_swarm_service.py#L636
https://github.com/hannseman/ansible/blob/4bfc254ac4dbf31afb0e5fc3d6c11c94bad9785e/lib/ansible/modules/cloud/docker/docker_swarm_service.py#L764

This means that not setting any value in the parameter `configs` will not try and send anything to docker daemons not supporting `configs` (API Version <1.30).

One issue that comes by setting `s.configs` to `None` (See line [#635](https://github.com/hannseman/ansible/blob/4bfc254ac4dbf31afb0e5fc3d6c11c94bad9785e/lib/ansible/modules/cloud/docker/docker_swarm_service.py#L635)) when falsy is that it will incorrectly report state as changed for clients supporting `configs`. This is because the docker client will return an empty list when no `configs` have been set and this is then compared to the nulled value. Since this issue is also present with other parameters (https://github.com/ansible/ansible/issues/49285) I think it should be solved in a separate PR. With this PR the `configs` parameter will at least be useable.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
docker_swarm_service

hannseman committed:
Fix crash when using configs parameter

felixfontein commented:
The standard approach for handling idempotence in such cases is to explicitly check for `None`. Replace
```.py
if self.configs != os.configs:
    differences.add('configs', parameter=self.configs, active=os.configs)
```
with
```.py
if self.configs is not None and self.configs != os.configs:
    differences.add('configs', parameter=self.configs, active=os.configs)
```
There's no need to handle this in another PR.

Also, you need a [changelog fragment](https://docs.ansible.com/ansible/latest/reference_appendices/release_and_maintenance.html#creating-new-fragments).

hannseman committed:
Don’t add difference for configs set to None

hannseman committed:
Add changelog fragment

felixfontein reviewed (approved):
LGTM

felixfontein commented:
@dariko @jwitko can you take a look at this one?

felixfontein commented:
@hannseman this PR fixes something which is only broken in `devel` currently, right? In that case, I guess we can remove the changelog fragment.

hannseman committed:
Revert "Add changelog fragment"

This reverts commit 8df497e3881d976e7417d2662a167694fbc33147.

hannseman commented:
@felixfontein yes you're right. I never got around to try that feature on 2.7 since the `docker_config`-module wasn't available. Reverted that commit.

hannseman commented:
Just now noticed some awkward behaviour of the `configs` options. When set to an empty list it will not be sent to docker. 

It's probably because of this line: https://github.com/ansible/ansible/blob/fd32760d7a1c42573e09055b822f0958f60a3874/lib/ansible/modules/cloud/docker/docker_swarm_service.py#L636

This check will be false if `ap.configs` is set to `[]` and thus `s.configs` will be `None`.

@felixfontein what do you think about changing this to `if ap['configs'] is not None`?

felixfontein commented:
One question: is it bad when it isn't send to docker? After all, if it's empty, the service doesn't need configs?

hannseman commented:
@felixfontein setting `configs` to an empty list could be used to clear out previously set configs. This is the behaviour of `networks`, `volumes` and `secrets`. In my opinion it will be confusing if `configs` behaves differently. 

@dariko yes but since but since the default of `module.params['configs']` is `None` my proposed solution will not send anything to docker if `configs` is not explicitly set to `[]`.

Like this:
```
        s.configs = None
        if ap['configs'] is not None:
            s.configs = []
            for param_m in ap['configs']:
                service_c = {}
                service_c['config_id'] = param_m['config_id']
                service_c['config_name'] = str(param_m['config_name'])
                service_c['filename'] = param_m.get('filename', service_c['config_name'])
                service_c['uid'] = int(param_m.get('uid', "0"))
                service_c['gid'] = int(param_m.get('gid', "0"))
                service_c['mode'] = param_m.get('mode', 0o444)
                s.configs.append(service_c)
```
https://github.com/ansible/ansible/blob/fd32760d7a1c42573e09055b822f0958f60a3874/lib/ansible/modules/cloud/docker/docker_swarm_service.py#L636

hannseman committed:
Enable config tests

hannseman committed:
Pass empty list of configs to docker

hannseman commented:
@dariko @felixfontein I've rebased this PR so we now have tests running for the `configs`-option. I also commited https://github.com/ansible/ansible/pull/50606/commits/008dbfdc710fed300426e25d50cf9a6dd251451c with my proposed changes to enable clearing `configs` by setting it to an empty list. Also added a testcase for this.

hannseman committed:
Properly clean up after all tests

felixfontein reviewed (approved):
I think it looks good. @dariko any opinions?

jwitko reviewed (approved):
shipit

felixfontein commented:
This needs to be rebased before it can be merged.

felixfontein reviewed (approved):
LGTM

felixfontein commented:
@hannseman thanks for the PR!
@dariko @jwitko thanks for reviewing!

Similar to #50882, I *think* this doesn't need to be backported, since the issue it fixes was caused by #45674, which wasn't backported. Please correct me if I'm wrong!
